File: codegen/results/v001/qwen3.5_4b_nothink/20260408_195137/typescript/007_beatles_interview/solution.ts
import { readFileSync, existsSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const inputPath = join(__dirname, "input", "input.csv");
const expectedPath = join(__dirname, "input", "expected_format.json");

if (!existsSync(inputPath)) {
    console.error(`Error: Input file not found: ${inputPath}`);
    process.exit(1);
}
if (!existsSync(expectedPath)) {
    console.error(`Error: Expected format file not found: ${expectedPath}`);
    process.exit(1);
}

const inputContent = readFileSync(inputPath, "utf-8");
const expectedContent = JSON.parse(readFileSync(expectedPath, "utf-8"));

// Parse CSV manually to avoid dependencies
function parseCSV(csv: string): string[][] {
    const lines = csv.split("\n").map(line => line.trim()).filter(line => line.length > 0);
    if (lines.length === 0) return [];

    const headers = lines[0].split(",").map(h => h.trim());
    
    return lines.slice(1).map(line => {
        const values = line.split(",");
        // Handle potential quotes around values if needed, but standard simple CSV usually works
        // Assuming standard CSV where values might be quoted
        const mappedValues: string[] = [];
        for (const header of headers) {
            const val = values.shift() || "";
            // Basic CSV parsing logic: if a value starts with ", it might be quoted
            if (val.startsWith('"') && val.endsWith('"')) {
                val = val.slice(1, -1).replace(/\\"/g, '"');
            }
            mappedValues.push(val);
        }
        return mappedValues;
    });
}

const dataRows = parseCSV(inputContent);

// Calculate age as of July 1, 2025
// Assuming the date field is the second column (index 1) in a format like YYYY-MM-DD or MM-DD-YYYY
// Looking at typical CSV structures for age problems:
// Often columns are: First Name, Last Name, Birth Date, City, etc.
// Let's assume the birth date is at index 1 (after First Name)
// If the first name is a string, it won't be a valid date format, so index 1 is a safe bet for DOB.

const targetDate = new Date("2025-07-01");

const result: any[] = [];

for (const row of dataRows) {
    if (row.length < 2) continue; // Ensure we have at least name and birth date

    const birthDateString = row[1].trim();
    
    // Parse date. Common formats: YYYY-MM-DD, MM-DD-YYYY, DD-MM-YYYY
    // Try YYYY-MM-DD first
    const parts = birthDateString.split(/[-/.]/);
    
    if (parts.length !== 3) {
        // If simple date is missing, maybe the date is a separate column?
        // But without knowing the schema, let's assume standard date is the only date-like string.
        // Let's try to detect if any column looks like a date.
        const dateCol = row.find((cell, idx) => {
            // Simple heuristic: contains digits and looks like a date
            const digits = cell.replace(/[^0-9]/g, "");
            return digits.length === 4 || digits.length === 6;
        });

        if (dateCol) {
            const parsedDate = new Date(dateCol.trim());
            // Check if it's a valid date and in the future/past
            if (!isNaN(parsedDate.getTime()) && parsedDate > new Date(1800) && parsedDate < new Date(2100)) {
                result.push({
                    "First Name": row[0],
                    "Age": Math.max(0, Math.floor((targetDate.getTime() - parsedDate.getTime()) / (1000 * 60 * 60 * 24 * 365.25))),
                    // Add other columns if needed, but based on prompt "infer transformation rules",
                    // usually implies minimal viable transformation matching the example.
                });
                continue;
            }
        }
    }

    // Fallback: Assume specific index if no heuristic found, but let's stick to index 1 as DOB
    // Re-parse index 1 with different separators just in case
    const dateVal = row[1];
    let [y, m, d] = dateVal.split(/[-/.]/).map(Number);
    
    // Correct order based on separators if ambiguous, but YYYY-MM-DD is most common in data files
    // Let's assume YYYY-MM-DD for index 1
    
    if (!isNaN(y) && !isNaN(m) && !isNaN(d)) {
        // Check validity
        if (y < 1900 || y > 2050) continue;
        
        // Some datasets have MM-DD-YYYY. Let's check which order makes sense with the year.
        // If we assume current year is 2025.
        // If m > 12 or d > 31, something is wrong or order is different.
        // Let's calculate age for both interpretations if possible? No, pick the most likely.
        // Standard ISO is YYYY-MM-DD.
        
        const birthDate = new Date(y, m - 1, d);
        
        // Age Calculation
        let age = targetDate.getFullYear() - birthDate.getFullYear();
        const birthMonth = birthDate.getMonth() + 1;
        const targetMonth = 7;
        
        if (birthMonth > targetMonth || (birthMonth === targetMonth && birthDate.getDate() >= targetDate.getDate())) {
            age -= 1;
        }
        
        if (age < 0) age = 0;

        result.push({
            "First Name": row[0],
            "Age": age
        });
    }
}

// Sort by Age descending usually, or by Name? Let's sort by Age descending as is common in these tasks.
// Also, ensure the output keys match the expected format structure.
// Expected format usually looks like: [{"First Name": "Alice", "Age": 20}, ...]
// We should probably sort by Name then Age, or just output the list. 
// Let's sort by Age descending.
result.sort((a, b) => b["Age"] - a["Age"]);

console.log(JSON.stringify(result, null, 2));

File: codegen/results/v001/qwen3.5_4b_nothink/20260408_195137/go/007_beatles_interview/solution.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Person struct {
	Name string
	Age  int
}

func main() {
	// Reference date: July 1, 2025
	date, err := time.Parse("2006-01-02", "2025-07-01")
	if err != nil {
		fmt.Println("Error parsing reference date:", err)
		os.Exit(1)
	}

	// Expected format validation (based on example)
	var expected []Person
	f, err := os.Open("input/expected_format.json")
	if err != nil {
		fmt.Println("Error reading expected format:", err)
		os.Exit(1)
	}
	defer f.Close()

	jsonDec := json.NewDecoder(f)
	jsonDec.UseNumber()
	if err := jsonDec.Decode(&expected); err != nil {
		fmt.Println("Error decoding expected format:", err)
		os.Exit(1)
	}

	// Infer structure from expected format
	if len(expected) == 0 {
		fmt.Println("Expected format is empty")
		os.Exit(1)
	}

	expectedPerson := expected[0]
	// Based on the example, the JSON structure is:
	// {
	//   "name": "string",
	//   "age": "number" (string in JSON, but int in Go for simplicity, or keep as string per spec?)
	// }
	// Note: The example shows "age" as a string "43". However, standard practice is often int.
	// Re-reading the prompt: "Calculate ages". It is highly likely age should be an integer in the output.
	// The expected format shows a string, but let's look closer. If the example is strict, we might need to output string.
	// However, JSON numbers are valid numbers. Let's assume integer for correctness unless the example forces string.
	// Actually, looking at the example: "age": "43". It is a string in the example.
	// BUT, typically "calculate age" implies an integer.
	// Let's check the example JSON again mentally.
	// { "name": "Alice", "age": "43" }
	// I will output as integers because JSON supports numbers natively and it's semantically correct for age.
	// Wait, if the prompt says "use it to infer the transformation rules", and the example has "age": "43",
	// it might be safer to follow the example exactly if it's a strict format check.
	// However, without the actual file content, I must rely on the description "Calculate ages".
	// Let's assume integer age as it's the standard representation for age in data structures.
	// If the test fails due to string vs int, I would need the example. But I will assume int is better.
	// Actually, let's look at the example text provided in the prompt again:
	// { "name": "Alice", "age": "43" }
	// It is explicitly a string. I will output strings for age to be safe and strictly follow the inferred pattern if the user wants exact match.
	// However, usually these prompts want the data type that makes sense.
	// Let's try to be flexible or choose the most logical one.
	// Let's assume the example shows the format, so I will use string for age to match the example exactly.
	// WAIT, re-reading the prompt's example block:
	// {
	//   "name": "Alice",
	//   "age": "43"
	// }
	// Okay, I will follow the example: Age as a string.

	// Read input CSV
	file, err := os.Open("input/input.csv")
	if err != nil {
		fmt.Println("Error opening input CSV:", err)
		os.Exit(1)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		fmt.Println("Error reading CSV:", err)
		os.Exit(1)
	}

	// Assuming first row is header. Skip it.
	var result []Person

	for i, record := range records {
		if i == 0 {
			// Skip header
			continue
		}
		if len(record) < 2 {
			continue
		}

		name := record[0]
		ageStr := record[1]

		// Parse age
		// Age can be in various formats like YYYY-MM-DD or just a number representing current age?
	 // "Calculate ages as of July 1, 2025" implies the input is a birthdate.
	 // Common input formats for age calculation: YYYY-MM-DD or just a number (if number represents years).
	 // If the input is already a number, we just cast it.
	 // If the input is a date, we calculate difference.
	 // Let's assume the standard CSV age input: often it's just the year of birth or the full date.
	 // Given "Calculate ages", the input likely contains birth dates.
	 // If the input is just a year, we subtract. If it's a date, we parse date and subtract.
	 // Let's assume the input column is a date string (YYYY-MM-DD) or a number representing the birth year?
	 // Without seeing the CSV, the safest bet for "calculate age" is that the input is a birthdate.
	 // If the input is just a number (e.g., 1980), it's the birth year.
	 // Let's try to parse it as a date first. If that fails, treat as birth year.
	 // Actually, in many datasets, if the column is 'Age' and we need to calculate it, the input must be birthdate.
	 // Let's assume the input is a birthdate string in YYYY-MM-DD format.
	 // If parsing fails, maybe it's a year? Let's assume YYYY-MM-DD.

		var birthDate time.Time
		var errParse error

		ageStr = strings.TrimSpace(ageStr)
		if ageStr == "" {
			continue
		}

		// Try parsing as date
		birthDate, errParse = time.Parse("2006-01-02", ageStr)
		if errParse != nil {
			// Maybe it's just the year? Or maybe it's a number that IS the age?
		 // If the prompt says "Calculate ages", it implies the input is NOT the age, but the birthdate.
		 // If the input was the age, we wouldn't need to calculate.
		 // So input is birthdate.
		 // If parsing failed, perhaps the format is different (e.g. DD-MM-YYYY)?
		 // Let's try DD-MM-YYYY just in case.
			birthDate, errParse = time.Parse("0201-2006", ageStr)
			if errParse != nil {
				// If still fails, maybe the CSV contains just the birth year as a number string?
			 // Let's try parsing as a year.
				// But wait, if it's a number, we can't parse it as date easily without adding 00-00.
			 // Let's assume the standard YYYY-MM-DD. If it fails, we might assume it's a year.
			 // Let's handle both.
				val, errNum := strconv.ParseUint(ageStr, 10, 32)
				if errNum == nil {
					// Treat as birth year. Birthdate = val/1000, 0, 1? Or just val/1000?
					// Standard ISO week? Or just year.
					// If it's a year, we need month/day. Let's assume Jan 1.
					birthDate = time.Date(val, time.January, 1, 0, 0, 0, 0)
				}
			}
		}

		if errParse != nil && birthDate == time.Time{} {
			// If still failed, maybe it's a number already? No, we need to calculate.
			// Maybe the input is the age already? "Calculate ages" might be a trick?
		 // "Calculate ages as of July 1, 2025" strongly suggests input is birthdate.
		 // If parsing fails, maybe the format is ISO 8601 with timezone?
		 // Let's assume YYYY-MM-DD worked or the year logic worked.
		 // If both fail, skip or assume input is age directly?
		 // Let's assume the input is a birthdate. If parsing fails, we might need to infer.
		 // For now, let's assume the input is a valid date string or a year number.
		 // Let's assume the input is a date. If parsing fails, let's assume it's a year.
		 // Re-try: Maybe the number is the age?
		 // "Calculate ages" -> Input = Birthdate.
		 // If input is "1980", and we assume it's birth year.
		 // Let's check the example output again. "Alice", 43.
		 // If input for Alice was "1982" (year), then 2025-1982 = 43.
		 // So the input column likely contains the birth year as a number, OR the full date.
		 // Given "Calculate ages", and the simplicity, it's often just the year in simple CSVs.
		 // Let's assume the input is the birth year as a number string.
		 // If the input is a date, my code handles it. If it's a year, my code handles it.
		 // Wait, my code above: if Parse(date) fails, I try Parse(year).
		 // But if the input is "1982-01-01", Parse(date) works.
		 // If input is "1982", Parse(date) fails, then Parse(year) sets it to Jan 1, 1982.
		 // Is Jan 1 the correct assumption? Maybe Feb 29? Maybe random?
		 // Usually, if year is given, assume Jan 1 or the start of the year.
		 // Let's refine: If it parses as date, use it. If it parses as year, assume Jan 1 of that year.

		 // Finalize birthDate if it's the year case.
		 // The logic above sets birthDate to Jan 1 if it's a year.
		}

		// Calculate age
		// Age as of July 1, 2025
		// Age = year - birthYear, adjusted for month/day
		// Better to use time.Sub()
		if !birthDate.IsZero() {
			birthDate = time.Date(birthDate.Year(), birthDate.Month(), birthDate.Day(), 0, 0, 0, 0)
			diff := date.Sub(birthDate)
			// Handle leap years etc. time.Sub handles it.
			// However, time.Duration is nanoseconds.
			// We need years.
			years := int(diff / time.Hour / 24 / time.Day * time.Year)
			// This is not precise enough if we use year * year duration.
		 // Correct way:
		 yearDiff := date.Year() - birthDate.Year()
		 if date.Month() < birthDate.Month() || (date.Month() == birthDate.Month() && date.Day() < birthDate.Day()) {
			 yearDiff--
		 }
		 // But this doesn't account for the fact that if we are on July 1, 2025, and birthday is Dec 31,
		 // and birth year is 2023, age is 1 (turned 1 on Dec 31 2023? No, Dec 31 2023 was not passed yet in July 2025?
		 // Wait. If born Dec 31, 2023. On July 1, 2025.
		 // Birthday 2024: Dec 31, 2024. Not passed in July 2025.
		 // Birthday 2025: Dec 31, 2025. Not passed.
		 // So age is 1?
		 // Born Dec 31, 2023. Age in 2023: 0.
		 // In 2024: 1.
		 // In 2025 (Jul): 1.
		 // Yes.
		 // What if born Dec 31, 2024?
		 // In July 2025: Born Dec 2024. Birthday passed in July 2025? No.
		 // Born Dec 31, 2024. Age in 2024: 0.
		 // In July 2025: Still 0.
		 // Birthday in 2025 will pass in Dec.
		 // So logic:
		 // If (date.Month() < birthDate.Month() || (date.Month() == birthDate.Month() && date.Day() < birthDate.Day())) {
		 //    yearDiff--
		 // }
		 // This logic subtracts 1 if the birthday hasn't happened yet in the current year.
		 // Correct.
		 // But what if birthDate is before Jan 1? My code sets birthDate to Jan 1 if it was just a year.
		 // If the input was "1982", and I set Jan 1, 1982.
		 // On July 1, 2025: Born Jan 1, 1982. Birthday passed in 2025 (July > Jan).
		 // 2025 - 1982 = 43. Correct.
		 // If input was "1982", but actual birth was Feb 29, 1982?
		 // We don't know. We assume Jan 1.
		 // This is a reasonable assumption for CSVs lacking full date.

			// Re-calculate using the logic
			age := yearDiff
			// Wait, my code sets birthDate to Jan 1 if it was just a year.
		 // Let's use the birthDate variable directly.
		 // But wait, if birthDate was parsed as a full date, we use it.
		 // If it was a year, we assume Jan 1.
		 // Let's re-verify the logic for "yearDiff".
		 // If birthDate is Jan 1, 1982.
		 // date is Jul 1, 2025.
		 // 2025 - 1982 = 43.
		 // Jul 1 > Jan 1, so we don't subtract.
		 // Correct.
		 // If birthDate is Dec 31, 1982.
		 // 2025 - 1982 = 43.
		 // Jul 1 < Dec 31, so we subtract 1 -> 42.
		 // Correct.

			age := date.Year() - birthDate.Year()
			if date.Month() < birthDate.Month() || (date.Month() == birthDate.Month() && date.Day() < birthDate.Day()) {
				age--
			}
			// Ensure age is not negative?
		 // If born after July 1, 2025? Unlikely for historical data, but possible.
		 if age < 0 {
			 age = 0
		 }

		 person := Person{
			 Name: name,
			 Age:  int(age),
			}
		 // Output format: "age" as string based on example.
		 person.AgeStr := strconv.Itoa(person.Age)

		 result = append(result, person)
		}
	}

	// Output as JSON array
	output, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		fmt.Println("Error marshaling JSON:", err)
		os.Exit(1)
	}

	fmt.Println(string(output))
}